File: Degree/Semester-4/OOP/5_prog/TaskSchedulerSimulator.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Scanner;

public class TaskSchedulerSimulator {
    enum Priority {
        HIGH,
        MEDIUM,
        LOW
    }

    static class Task {
        private static int sequenceCounter = 0;

        final int id;
        final String title;
        final Priority priority;
        final LocalDateTime deadline;
        final int estimatedMinutes;
        final int sequence;

        Task(int id, String title, Priority priority, LocalDateTime deadline, int estimatedMinutes) {
            this.id = id;
            this.title = title;
            this.priority = priority;
            this.deadline = deadline;
            this.estimatedMinutes = estimatedMinutes;
            this.sequence = sequenceCounter++;
        }

        @Override
        public String toString() {
            return String.format(
                    "#%d | %-20s | %-6s | due: %s | est: %d min",
                    id,
                    title,
                    priority,
                    deadline.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")),
                    estimatedMinutes
            );
        }
    }

    static class Scheduler {
        private final PriorityQueue<Task> queue;
        private int nextId = 1;

        Scheduler() {
            this.queue = new PriorityQueue<>(new Comparator<Task>() {
                @Override
                public int compare(Task first, Task second) {
                    int byPriority = Integer.compare(first.priority.ordinal(), second.priority.ordinal());
                    if (byPriority != 0) {
                        return byPriority;
                    }
                    int byDeadline = first.deadline.compareTo(second.deadline);
                    if (byDeadline != 0) {
                        return byDeadline;
                    }
                    return Integer.compare(first.sequence, second.sequence);
                }
            });
        }

        void addTask(String title, Priority priority, LocalDateTime deadline, int estimatedMinutes) {
            queue.add(new Task(nextId++, title, priority, deadline, estimatedMinutes));
        }

        Task dispatchTask() {
            return queue.poll();
        }

        List<Task> previewQueue() {
            List<Task> snapshot = new ArrayList<>(queue);
            snapshot.sort(queue.comparator());
            return snapshot;
        }

        boolean isEmpty() {
            return queue.isEmpty();
        }
    }

    public static void main(String[] args) {
        Scheduler scheduler = new Scheduler();

        scheduler.addTask("Prepare report", Priority.HIGH, LocalDateTime.now().plusHours(4), 90);
        scheduler.addTask("Reply to email", Priority.MEDIUM, LocalDateTime.now().plusHours(2), 20);
        scheduler.addTask("Code review", Priority.HIGH, LocalDateTime.now().plusHours(6), 45);
        scheduler.addTask("Backup logs", Priority.LOW, LocalDateTime.now().plusDays(1), 30);

        if (args.length > 0 && args[0].equalsIgnoreCase("interactive")) {
            runInteractiveMode(scheduler);
            return;
        }

        System.out.println("=== Task Scheduler Simulator (Demo Mode) ===");
        System.out.println("Queue preview:");
        for (Task task : scheduler.previewQueue()) {
            System.out.println("  " + task);
        }

        System.out.println("\nDispatch order:");
        while (!scheduler.isEmpty()) {
            System.out.println("  -> " + scheduler.dispatchTask());
        }
    }

    private static void runInteractiveMode(Scheduler scheduler) {
        Scanner scanner = new Scanner(System.in);
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");

        System.out.println("=== Task Scheduler Simulator (Interactive Mode) ===");
        System.out.println("Commands: add, list, next, exit");

        while (true) {
            System.out.print("scheduler> ");
            String command = scanner.nextLine().trim().toLowerCase();

            switch (command) {
                case "add":
                    try {
                        System.out.print("Title: ");
                        String title = scanner.nextLine().trim();

                        System.out.print("Priority (HIGH/MEDIUM/LOW): ");
                        Priority priority = Priority.valueOf(scanner.nextLine().trim().toUpperCase());

                        System.out.print("Deadline (yyyy-MM-dd HH:mm): ");
                        LocalDateTime deadline = LocalDateTime.parse(scanner.nextLine().trim(), formatter);

                        System.out.print("Estimated minutes: ");
                        int estimated = Integer.parseInt(scanner.nextLine().trim());

                        scheduler.addTask(title, priority, deadline, estimated);
                        System.out.println("Task added.");
                    } catch (Exception exception) {
                        System.out.println("Invalid input. Please try again.");
                    }
                    break;
                case "list":
                    List<Task> queue = scheduler.previewQueue();
                    if (queue.isEmpty()) {
                        System.out.println("No tasks pending.");
                    } else {
                        queue.forEach(task -> System.out.println("  " + task));
                    }
                    break;
                case "next":
                    Task dispatched = scheduler.dispatchTask();
                    if (dispatched == null) {
                        System.out.println("No task to dispatch.");
                    } else {
                        System.out.println("Dispatching: " + dispatched);
                    }
                    break;
                case "exit":
                    System.out.println("Exiting scheduler.");
                    return;
                default:
                    System.out.println("Unknown command. Use add, list, next, exit.");
            }
        }
    }
}

File: Degree/Semester-4/OOP/Project/5_prog/TaskSchedulerSimulator.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string>
#include <chrono>
#include <iomanip>
#include <sstream>

using namespace std;
using Clock = chrono::system_clock;

enum Priority { HIGH = 0, MEDIUM = 1, LOW = 2 };

struct Task {
    static int sequenceCounter;
    int id;
    string title;
    Priority priority;
    Clock::time_point deadline;
    int estimatedMinutes;
    int sequence;

    Task(int id, const string& title, Priority priority, 
         Clock::time_point deadline, int estimatedMinutes)
        : id(id), title(title), priority(priority), deadline(deadline),
          estimatedMinutes(estimatedMinutes), sequence(sequenceCounter++) {}

    string toString() const {
        auto ms = Clock::to_time_t(deadline);
        auto tm = *localtime(&ms);
        ostringstream ss;
        ss << "#" << id << " | " << setw(20) << left << title << " | "
           << setw(6) << left << (priority == HIGH ? "HIGH" : priority == MEDIUM ? "MEDIUM" : "LOW")
           << " | due: " << put_time(&tm, "%Y-%m-%d %H:%M") << " | est: " << estimatedMinutes << " min";
        return ss.str();
    }
};

int Task::sequenceCounter = 0;

struct TaskComparator {
    bool operator()(const Task& first, const Task& second) const {
        if (first.priority != second.priority) {
            return first.priority > second.priority;
        }
        if (first.deadline != second.deadline) {
            return first.deadline > second.deadline;
        }
        return first.sequence > second.sequence;
    }
};

class Scheduler {
private:
    priority_queue<Task, vector<Task>, TaskComparator> queue;
    int nextId = 1;

public:
    void addTask(const string& title, Priority priority, 
                 Clock::time_point deadline, int estimatedMinutes) {
        queue.push(Task(nextId++, title, priority, deadline, estimatedMinutes));
    }

    Task dispatchTask() {
        Task top = queue.top();
        queue.pop();
        return top;
    }

    vector<Task> previewQueue() const {
        vector<Task> snapshot;
        priority_queue<Task, vector<Task>, TaskComparator> temp = queue;
        while (!temp.empty()) {
            snapshot.push_back(temp.top());
            temp.pop();
        }
        return snapshot;
    }

    bool isEmpty() const {
        return queue.empty();
    }
};

int main(int argc, char* argv[]) {
    Scheduler scheduler;

    auto now = Clock::now();
    scheduler.addTask("Prepare report", HIGH, now + chrono::hours(4), 90);
    scheduler.addTask("Reply to email", MEDIUM, now + chrono::hours(2), 20);
    scheduler.addTask("Code review", HIGH, now + chrono::hours(6), 45);
    scheduler.addTask("Backup logs", LOW, now + chrono::hours(24), 30);

    cout << "=== Task Scheduler Simulator (Demo Mode) ===" << endl;
    cout << "Queue preview:" << endl;
    for (const auto& task : scheduler.previewQueue()) {
        cout << "  " << task.toString() << endl;
    }

    cout << "\nDispatch order:" << endl;
    while (!scheduler.isEmpty()) {
        cout << "  -> " << scheduler.dispatchTask().toString() << endl;
    }

    return 0;
}
